File: python/295-Find Median from Data Stream.py
from sortedcontainers import SortedList

def get_median_from_sorted(sorted_nums):
    n = len(sorted_nums)
    if n % 2 == 0:
        median = (sorted_nums[(n-1)//2] + sorted_nums[(n-1)//2+1]) / 2
    else:
        median = sorted_nums[(n-1)//2]
    return median
# class MedianFinder:
#     "Sort"
#     def __init__(self):
#         "Space: O(n)"
#         self.nums = []
#     def addNum(self, num: int) -> None:
#         "Time: O(nlogn)"
#         self.nums.append(num)
#         self.nums.sort()
#     def findMedian(self) -> float:
#         "Time: O(1)"
#         return get_median_from_sorted(self.nums)
# class MedianFinder:
#     "Insertion"
#     def __init__(self):
#         "Space: O(n)"
#         self.nums = []
#     def addNum(self, num: int) -> None:
#         "Time: O(n)"
#         self.nums.append(float("+inf"))
#         n = len(self.nums)
#         for i in range(n):
#             if self.nums[i] >= num:
#                 self.nums.insert(i, num)
#                 break
#         del self.nums[-1]
#     def findMedian(self) -> float:
#         "Time: O(1)"
#         return get_median_from_sorted(self.nums)
class MedianFinder:
    "SortedList or Balanced Binary Search Tree"
    def __init__(self):
        "Space: O(n)"
        self.nums = SortedList()
    def addNum(self, num: int) -> None:
        "Time: O(logn)"
        self.nums.add(num)
    def findMedian(self) -> float:
        "Time: O(1)"
        return get_median_from_sorted(self.nums)

median_finder = MedianFinder()
median_finder.addNum(1)
median_finder.addNum(2)
print(median_finder.findMedian()) # 1.5
median_finder.addNum(3) 
print(median_finder.findMedian()) # 2

median_finder = MedianFinder()
median_finder.addNum(-1)
median_finder.addNum(-2)
print(median_finder.findMedian()) # -1.5

File: cpp/295. Find Median from Data Stream.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;
typedef set<pi>::iterator si;

/*
https://www.youtube.com/watch?v=60xnYZ21Ir0
*/

// class MedianFinder {
// // Balanced Binary Search Tree
// public:
//     // Space: O(n)
//     set<pi> s; // {(value, index)}
//     si mIter;
//     MedianFinder() {
//     }
//     void addNum(int num) {
//         // Time: O(logn)
//         pi curElem = make_pair(num, s.size());
//         auto [iter, _] = s.insert(curElem);
//         if(s.size() == 1) { // First one
//             mIter = iter;
//             return;
//         }
//         if(s.size()%2) { // odd, after insert
//             if(curElem > *mIter) mIter++;
//         } else { // even, after insert
//             if(curElem < *mIter) mIter--;
//         }
//     }
//     double findMedian() {
//         // Time: O(1)
//         if(s.size()%2) return (*mIter).first; // odd
//         else return ( (double)(*mIter).first + (*next(mIter)).first ) / 2; // even
//     }
// };

class MedianFinder {
// Less Heap + Larger Heap
public:
    // Space: O(n)
    priority_queue<int> less; // Max heap
    priority_queue<int, vi, greater<int>> larger; // Min heap
    MedianFinder() {
    }
    void addNum(int num) {
        // Time: O(logn)
        if(less.size()==0 || num <= less.top()) less.push(num);
        else larger.push(num);

        if(larger.size() > less.size()) {
            less.push(larger.top());
            larger.pop();
        } 
        if(less.size() >= larger.size()+2) {
            larger.push(less.top());
            less.pop();
        }
    }
    double findMedian() {
        // Time: O(1)
        if( (less.size()+larger.size()) % 2 ) return less.top(); // odd
        else return ( (double)less.top() + larger.top() ) / 2; // even
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    MedianFinder mf;

    mf.addNum(1);
    mf.addNum(2);
    cout << mf.findMedian() << endl; // -> 1.5
    mf.addNum(3); 
    cout << mf.findMedian() << endl; // -> 2
}